crates/typst/src/eval/version.rs

```diff
@@ -0,0 +1,126 @@
+use std::cmp::Ordering;
+use std::fmt::{Debug, Display, Formatter};
+use std::hash::Hash;
+use std::iter::repeat;
+
+use ecow::EcoVec;
+
+use super::{Array, Value};
+use crate::diag::{bail, StrResult};
+
+/// A version, with any number of components.
+///
+/// The list of components is semantically extended
+/// by an infinite list of zeros.
+/// This means that, for example, `0.8` is the same as `0.8.0`.
+/// As a special case, the empty version (that has no components at all)
+/// is the same as `0`, `0.0`, `0.0.0`, and so on.
+// reason: hash is for incremental compilation, so it needs to be different
+// for values that display differently.
+// It being different from `Eq` is consistent with many other typst types.
+#[allow(clippy::derived_hash_with_manual_eq)]
+#[derive(Default, Clone, Hash)]
+pub struct Version(EcoVec<u32>);
+
+impl Version {
+    pub const COMPONENT_NAMES: [&'static str; 3] = ["major", "minor", "patch"];
+
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    fn get(&self, index: usize) -> Option<u32> {
+        self.0.get(index).copied()
+    }
+
+    /// Get a component of a version.
+    ///
+    /// Always non-negative. Returns `0` if the version isn't specified to the necessary length.
+    pub fn at(&self, index: i64) -> StrResult<i64> {
```

Comment:
at generally also allows indexing from the back

Reply:
Given that a version is theoretically infinite, i don't know what would be the right thing to do here...

Reply:
Is it really conceptually infinite? I think it has pretty clearly a number of specified components. It's only that it compares equivalently to any zero-extended variant of it. So, indexing from the back would just index from the back into the internal vector.

Reply:
Hm, I guess that's actually the only reasonable thing to do with negative indices (besides just rejecting them outright) so there's probably no harm in just doing it.